testing/code/test_source.py

```diff
@@ -623,6 +623,28 @@ def test_comment_in_statement() -> None:
         )
 
 
+def test_source_with_decorator() -> None:
+    """Test behavior with Source / Code().source with regard to decorators."""
+    from _pytest.compat import get_real_func
+
+    @pytest.mark.foo
+    def deco_mark():
+        assert False
+
+    src = inspect.getsource(deco_mark)
+    assert str(Source(deco_mark, deindent=False)) == src
+    assert src.startswith("    @pytest.mark.foo")
+
+    @pytest.fixture
+    def deco_fixture():
+        assert False
+
+    src = inspect.getsource(deco_fixture)
```

Comment:
Sorry, I should have been clearer, what I meant was something along the lines:

```suggestion
    # currenly getsource does not unwrap decorators, testing the
    # existing behavior here for explicitness, but perhaps we should revisit/change this
    # in the future
    src = inspect.getsource(deco_fixture)
```

My thinking is that someone might want to make `inspect.getsource` unwrap decorators, but then this test will fail and they will be confused thinking this is intentional behavior, rather than accidental.

Reply:
Well, it is intentional / the current behavior - and automatically unwrapping would be a breaking change AFAICT, which requires a new argument to control it.  pytest currently likely unwraps things before (at least sometimes ;)).

As for your suggested change: `inspect.getsource` _does_ unwrap it - so it would only add to the confusion.  I'll extend the docstring.

Reply:
```diff
diff --git i/testing/code/test_source.py w/testing/code/test_source.py
index 3d5e4c142..1119e4a53 100644
--- i/testing/code/test_source.py
+++ w/testing/code/test_source.py
@@ -624,7 +624,11 @@ def test_comment_in_statement() -> None:


 def test_source_with_decorator() -> None:
-    """Test behavior with Source / Code().source with regard to decorators."""
+    """Test behavior with `Source` (also used by `Code().source`)
+    with regard to decorators.
+
+    Unlike `inspect.getsource()` it does not unwrap them, for which
+    `get_real_func()` can be used."""
     from _pytest.compat import get_real_func

     @pytest.mark.foo
```
But I really do not like the extra verbosity etc - I do not think it helps.
A lot of tests are about current behavior and it is clear that their output needs to be changed.

Reply:
Oh gosh I'm so sorry, I meant for that comment to be before line 644 down below 🤦‍♂ 

It is supposed to be about the behavior of `str(Source())`, not about `inspect.getsource`, really sorry about the confusion.

`inspect.getsource` does the correct unwrapping that we expect, while `Source` doesn't. `Source` not unwrapping I believe is accidental, and that's why I wanted to leave a comment there stating that.

Reply:
Ok.  No rush - and if is really expected to unwrap we could also fix it / improve it via here instead.

Reply:
But I still think a comment there would help avoid confusion in the future:

```python
# Source() does not unwrap functions at this time, but it probably should
assert str(Source(deco_fixture)).startswith("@functools.wraps(function)")
assert str(Source(get_real_func(deco_fixture), deindent=False)) == src
```

What do you think?

Reply:
1. if it gets fixed/changed the comment is void/wrong
2. it it gets fixed, would you still want a comment "Source() unwraps functions just like inspect.getsource()" or similar?  My (repeated) point here is that tests in general do not have comments à la "this tests the current behavior, it might need to be adjusted when changing/fixing things".

Anyway, I've said that this PR could be extended to fix it, so there is really no reason to have a comment inbetween, and you should rather have given your opinion on if this should be fixed directy (instead of indirectly via a suggested code comment ("but it probably should")).

Reply:
1. Yes.
2. If it gets fixed in the future, then we remove the comment. The comment is there so "future generations" don't have to debate if this is intentional or not, which is exactly what we have been doing.

> Anyway, I've said that this PR could be extended to fix it

I'm OK with merging as is, adding a comment. I'm also OK with fixing it by using an option, and then of course not adding a comment.

I don't have any immediate need to change it myself, and because as you have often said "it is out of scope of this PR", my reaction here is to just ask for the minimum of changes, instead of piling more requests on it.

But if you want to do some extra work here and change this, then by all means, I believe the change is good (as I have mentioned before).

> you should rather have given your opinion on if this should be fixed directy (instead of indirectly via a suggested code comment ("but it probably should")).

Just to clarify, I don't think this needs fixing immediately: it is used only internally, users shouldn't be using this class at all. If we have the internal need for this functionality now, then definitely it is worth doing it.